feat(blog): Customizing Styles in Gatsby Themes with Theme UI (#15290)
gatsbyjs/gatsby: docs/blog/2019-07-03-customizing-styles-in-gatsby-themes-with-theme-ui/index.md

date: 2019-07-03
author: Brent Jackson
tags:
  - themes
  - css
  - styles
  - theme-ui
---

With Gatsby themes we want to make the experience of building a site as flexible and extensible as possible.
Themes allow you to reuse site configurations, sets of plugins, and components across multiple Gatsby sites.
We've tried to design the API of themes to [_progressively disclose complexity_][progressive disclosure],
where smart defaults are wrapped into abstractions, but customization and configuration are always available at your fingertips.
The styles used in our official themes are no exception.

## Say hello to our official themes

Today, we're releasing two official themes: a blog theme inspired by Dan Abromov's [overreacted.io][], and our notes theme,
which is intended for freeform, non-linear content and inspired by Tom Critchlow's concept of a _[Digital Garden][]_.
Both themes include their own styles so that you can build and publish a blog without starting from scratch.
In order to make these themes truly _themeable_, the values used for typography, color, layout, and other design tokens
can be customized by editing part or all of a `theme` configuration object.

> This also means you can create and publish your own theme based on the official one,

Reviewer: Random question: do these line breaks after the end of a sentence create a new paragraph in Markdown? (I wanted to make sure the post wasn't full of `<span>`s but I'm hoping to avoid pulling it up locally.)

Author: Line breaks like this *shouldn't* cause weird markup or breaks. The one major exception is GitHub-flavored in GitHub Issues, which does have problems with this, but I can double check the rendered markup locally

Author: It *does* appear to render as a single paragraph locally

> with custom typography and colors, by editing a few plain objects instead of recreating entire components.

Author: I don't have a CS background so I might be wrong here, and I definitely would like this to be as clear as possible, but I think of JSON as the language-agnostic syntax, whereas *plain object* was meant as a reference to Plain Old JavaScript Object (POJO) – i.e. not a function or other object type. We do want these to be JSON-serializable, but that's not a hard requirement. Glancing at both the Emotion and Styled Components docs, they simply refer to `theme` as an object. 

The wording does sound awkward here though, do you think something like this would be better?
```suggestion
with custom typography and colors, by editing an object instead of recreating entire components.
```

> with custom typography and colors, by editing a few plain objects instead of recreating entire components.

Reviewer: I think `an object` sounds fine!

For sites that have multiple themes installed,
and to help ensure a consistent look and feel,
you can choose which theme's styles take precedence based on their order in your Gatsby configuration.
The styles provided by each theme are only applied to the pages it generates,
giving you complete control of your site's styles.

Huge thanks to our own Amberley Romo & John Otander for their incredible work on these themes.

## Theme UI

To enable this level of style customization,
we've been working on a new open source library called _[Theme UI][]_.
For authors of Gatsby themes, this library is entirely optional and in no way coupled to Gatsby themes,
but we encourage you to check it out.

> Gatsby themes can be built without Theme UI, and Theme UI can be used without Gatsby.

Reviewer: Love this level of detail! It really helps to answer these questions here. 🙌

Author: Thanks! I'm sort of anticipating some questions like this, and I don't want to scare anyone off 🙂

Styled System is great for creating a custom design system in React,
but it falls short when it comes to styling content-heavy sites and doesn't provide much guidance for creating truly themeable UI.
Styled System also takes some effort to set up and generally requires creating custom components to get started.

Theme UI is intended to address some of these issues by:

- Providing guidelines for how to create themeable UI and making configuration more portable
- Allowing developers to use consistent styles anywhere in React with minimum effort
- Enabling a simple way to style [MDX][] content, while keeping styles isolated
- Abstracting some of the implementation details so you don't need to start from scratch

Theme UI currently includes optional packages for integrating with other libraries like [Typography.js][],
and we plan to build more tools and integrations around the core foundation of Theme UI to make styling Gatsby sites even better in the future.

## Customizing the blog theme

To quickly demonstrate how this works,
we'll walk through customizing the colors and typography used in the blog theme.

> `todo: basic setup instructions from theme starter`

Reviewer: This could be:

```gatsby new my-blog https://github.com/gatsbyjs/gatsby-starter-blog-theme```

Once you've installed the blog theme and added it to the `plugins` array in your `gatsby-config.js`,
add a `src/gatsby-plugin-theme-ui` directory to your site.

```shell
mkdir src/gatsby-plugin-theme-ui
```

> Create an `index.js` file in this directory to export the theme configuration object.

Reviewer: ```suggestion
Create an `index.js` file in this new directory to export the theme configuration object.
```
